### PR TITLE
fix(auth0): preserve old primary email before switching primary

### PR DESCRIPTION
**Pull request by fayazg**
## Summary

- Preserve the previous primary email before switching the primary in the Auth0 adapter, so it stays reachable — matching the mock adapter and the documented behavior.
- Preservation is conditional: it is **skipped** when the old primary is already backed by a sufficient identity for that address — a verified `email`/OTP identity **or** a Google OAuth (`google-oauth2`) login (Google already covers the address). It runs for everything else: a non-Google provider only (LinkedIn, GitHub, enterprise, …) or no backing identity at all.
- When it runs, the old primary is materialized as a verified, **user-removable** `email`-connection identity (not system-managed), so it is listed as an alternate and re-promotable, and the user can later unlink it.
- Fail-loud: preservation runs first; on failure the just-created stub is rolled back and the switch is aborted, never silently dropping the old primary.
- Refactor: extract `createAndLinkEmailIdentity`, shared by `AddSystemManagedEmail` (system-managed) and the preservation path (user-removable); add `deleteEmailConnectionStub` for connection-guarded rollback; add a `GoogleOAuth2Connection` constant.
- Add an optional `Transport` seam to `pkg/httpclient` so the Auth0 Management API flow can be unit-tested.
- Docs: update `docs/subjects/user_emails.md` and the Auth0 README "Set Primary Email" section.

## Test plan

- [x] `go build ./...`
- [x] `go test ./...` (full suite green)
- [x] `golangci-lint run` — 0 new issues vs main
- [x] `SetPrimaryEmail`: old primary with no backing identity → create+link (verified, no `app_metadata`) then PATCH, in order
- [x] `SetPrimaryEmail`: old primary backed by Google OAuth → NOT re-created (just PATCH)
- [x] `SetPrimaryEmail`: old primary backed only by a non-Google social identity (LinkedIn) → preserved as an `email` identity
- [x] `SetPrimaryEmail`: old primary already a verified email identity / equals new email → no stub created
- [x] `SetPrimaryEmail`: unverified or unknown new email → rejected before any mutation
- [x] `SetPrimaryEmail`: preservation failure → rollback stub, abort (no PATCH)
- [x] `AddSystemManagedEmail` still creates a system-managed stub; 409 maps to a validation error
- [x] Mock adapter `SetPrimaryEmail` reference behavior

Fixes LFXV2-2256

🤖 Generated with Claude Code

### CODE REVIEW

File: internal/infrastructure/auth0/user.go [hunk 1]
```diff
@@ -676,6 +708,31 @@ type setPrimaryEmailRequest struct {
...
+// primary must be preserved as a verified email identity before it is replaced.
+// Matching is case-insensitive on the identity's email.
+func hasSufficientPrimaryEmailIdentity(user *model.User, email string) bool {
+	for _, id := range user.Identities {
+		if !strings.EqualFold(id.Email, email) {
+			continue
+		}
+		if id.Connection == constants.EmailConnection || id.Connection == constants.GoogleOAuth2Connection {
+			return true
+		}
```
[the remark this answers is not in the document]
Comment (fayazg) [reply internal/infrastructure/auth0/user.go]: Good catch — fixed in 942ea3b. `hasSufficientPrimaryEmailIdentity` now requires `email_verified` for an `email`-connection identity to count as sufficient; Google OAuth stays unconditional (Google verifies the address). An unverified email primary now triggers preservation, matching the "verified email/OTP" wording. Added a regression subtest for the unverified case.

[the remark this answers is not in the document]
Comment (fayazg) [reply internal/infrastructure/auth0/user.go:739]: Agreed — applied in 942ea3b. The check is now `id.Connection == constants.EmailConnection && id.EmailVerified` for email, with Google OAuth kept unconditional. Added a subtest asserting an unverified email primary is preserved.

File: internal/infrastructure/auth0/user_test.go [hunk 2]
```diff
@@ -889,3 +892,336 @@ func containsSubstring(s, substr string) bool {
...
+	}
+}
+
+func (f *fakeAuth0Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	var bodyBytes []byte
+	if req.Body != nil {
+		bodyBytes, _ = io.ReadAll(req.Body)
+		_ = req.Body.Close()
+	}
+	f.calls = append(f.calls, recordedCall{method: req.Method, path: req.URL.Path, body: string(bodyBytes)})
```
[the remark this answers is not in the document]
Comment (fayazg) [reply internal/infrastructure/auth0/user_test.go:950]: Thanks for flagging — I dug into this and it's actually a non-issue. `net/http` stores the **decoded** path in `req.URL.Path` (it decodes `%7C` back to `|`), and the fake reads `req.URL.Path`, so it sees `auth0|test123`, matching the fixtures. That's why the full suite passes with decoded IDs. (The escaped form lives in `req.URL.RawPath`/`EscapedPath()`, which the fake doesn't use.)

File: docs/subjects/user_emails.md [hunk 3]
```diff
@@ -209,7 +209,7 @@ nats request lfx.auth-service.user_emails.set_primary \
 - Requires the `update:current_user_identities` scope in the JWT
 - The target email must already exist as a verified alternate email on the user's account; it cannot be an arbitrary new address
 - The email input is normalized to lowercase before processing; matching is case-insensitive
-- After a successful update, the previous primary email becomes an alternate email on the account
+- After a successful update, the previous primary email is kept reachable. Unless it is already backed by a verified email/OTP identity or a Google login (which already covers the address), it is preserved as a verified, **user-removable** alternate identity (not system-managed) so the user can later unlink it like any other alternate email. A previous primary backed only by a non-Google provider (e.g. LinkedIn) is preserved this way too. Preservation happens before the switch: if the previous primary cannot be preserved, the switch is aborted and the account is left unchanged rather than dropping the old primary
 - For detailed Auth0-specific behavior, see: [`../internal/infrastructure/auth0/README.md`](../internal/infrastructure/auth0/README.md)
```
[the remark this answers is not in the document]
Comment (fayazg) [reply docs/subjects/user_emails.md]: Fixed in 942ea3b — updated to `../../internal/infrastructure/auth0/README.md` to match the correct depth already used on line 119. Good catch.

[the remark this answers is not in the document]
Comment (fayazg) [reply docs/subjects/user_emails.md]: Fixed in 942ea3b — link now uses `../../internal/infrastructure/auth0/README.md`, matching the depth on line 119.

File: internal/infrastructure/auth0/user.go [hunk 4]
```diff
@@ -715,6 +779,26 @@ func (u *userReaderWriter) SetPrimaryEmail(ctx context.Context, userID string, e
...
+	// email/OTP login unless it is already backed by a verified email identity or
+	// a Google login. So unless the old primary is sufficiently backed (see
+	// hasSufficientPrimaryEmailIdentity), create+link it as a normal,
+	// user-removable verified email identity first. A primary backed only by a
+	// non-Google provider (LinkedIn, GitHub, enterprise, etc.) is preserved this
+	// way too. Done first so that any failure leaves the account unchanged rather
+	// than silently dropping the old primary.
+	oldPrimary := fullUser.PrimaryEmail
+	if oldPrimary != "" && !strings.EqualFold(oldPrimary, email) && !hasSufficientPrimaryEmailIdentity(fullUser, oldPrimary) {
+		if _, errPreserve := u.createAndLinkEmailIdentity(ctx, userID, oldPrimary, nil); errPreserve != nil {
```
Comment (audigregorie) [inline internal/infrastructure/auth0/user.go:794]: **Major — preserve path aborts the switch when the old primary has an *unverified* email identity**

When `oldPrimary` is backed by an unverified `email`-connection identity, `hasSufficientPrimaryEmailIdentity` returns `false` (it requires `EmailVerified`), so this calls `createAndLinkEmailIdentity(ctx, userID, oldPrimary, nil)`. But `oldPrimary` already exists on the `email` connection, so Auth0's `POST /api/v2/users` returns **409**, which `createAndLinkEmailIdentity` maps to `errors.NewValidation("email already linked")`. That error propagates here and **aborts the entire `SetPrimaryEmail`** — a user whose old primary carries an unverified email identity can never switch primary off it.

The verified-only rule is correct for *login usability*, but wrong as a *create-or-not* gate. Consider short-circuiting preservation when **any** `email`-connection identity for `oldPrimary` already exists (verified or not), or making the preserve call treat a 409 as idempotent success rather than fatal.

(See also the test at `user_test.go:1129`, which currently green-lights this broken path because the fake transport always returns 201 on create.)

File: internal/infrastructure/auth0/user_test.go [hunk 5]
```diff
@@ -889,3 +892,355 @@ func containsSubstring(s, substr string) bool {
...
+			`{"connection":"email","provider":"email","profileData":{"email":"new@example.com","email_verified":true}},` +
+			`{"connection":"email","provider":"email","profileData":{"email":"old@example.com","email_verified":false}}]}`
+		ft := newFakeAuth0(testPrimaryUserID, getUser)
+		rw := newTestReaderWriter(ft)
+
+		err := rw.SetPrimaryEmail(ctx, testPrimaryUserID, "new@example.com")
+		require.NoError(t, err)
+
+		// An unverified email identity does not count as sufficient, so preservation runs.
+		assert.Equal(t, 1, ft.countFor(http.MethodPost, "/api/v2/users"), "unverified email primary should be preserved")
```
Comment (audigregorie) [inline internal/infrastructure/auth0/user_test.go]: **Minor — this test masks the unverified-email edge case**

The fake transport always returns `201` for `POST /api/v2/users`, so asserting `NoError` + one create here green-lights exactly the scenario that returns **409** against real Auth0 (see the `user.go` preserve-path comment). Against the live API this case aborts the switch with `"email already linked"`.

Once the 409/idempotency handling is decided, drive this case with `ft.createStatus = http.StatusConflict` and assert the intended outcome (switch proceeds, or a clear, deliberate error) instead of a silent success.

**Review by audigregorie — COMMENTED** [review]
## Code Review Summary

Solid, well-documented change with real HTTP-flow test coverage and a fail-loud preserve-then-promote design. Build + `go test ./internal/infrastructure/auth0/...` are green. One edge case (old primary backed by an *unverified* email identity) is mishandled and the new test masks it — details inline.

### What's done well

- Fail-loud ordering: preservation runs before the PATCH with best-effort rollback, so failure leaves the account intact.
- `deleteEmailConnectionStub` is well-guarded (refuses non-`email` identities and the 0-identity case).
- The `Transport` test seam adds genuine HTTP-flow coverage (call ordering, payloads, rollback) that didn't exist before.

File: internal/infrastructure/auth0/user.go (hunk 4 above)
Comment (fayazg) [reply internal/infrastructure/auth0/user.go:794]: Great catch @audigregorie — fixed in 5411196. You're right that the verified-only check was wrong as a create-or-not gate. `hasSufficientPrimaryEmailIdentity` now treats **any** `email`-connection identity (verified or not) as sufficient — it's OTP-reachable and Auth0 self-verifies it on the next login, and a duplicate can't be created — so the unverified-email case now skips preservation and the switch proceeds instead of hitting the 409 abort. Google is now sufficient **only when verified**; an unverified Google identity (or a non-Google provider, or no backing identity) still materializes a verified email identity.

File: internal/infrastructure/auth0/user_test.go (hunk 5 above)
Comment (fayazg) [reply internal/infrastructure/auth0/user_test.go]: Fixed in 5411196. The unverified-email subtest now asserts the switch is skipped (GET + PATCH only, no create), reflecting the corrected gate. I also added a separate case driving an **unverified Google** primary, which asserts preservation does run (create + link + PATCH) — covering the branch that legitimately materializes a verified email identity.

**Review by audigregorie — APPROVED** [review]
